File: packages/ledgerbench/src/report.ts
/**
 * The published run artifact.
 *
 * Everything the specification requires to travel with a number travels in this
 * object: dataset version, fixture version, seed, code version, policy version, gate
 * backend and model hash, plus the raw per-fixture rows the metrics were computed
 * from. A report that cannot be traced back to the rows that produced it is an
 * advertisement, not a benchmark.
 *
 * `report.ts` in `python/evals` consumes this shape and is the only thing that
 * renders a published table.
 */
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { DEFAULT_COMMIT_POLICY } from "@veritymem/contracts";
import { loadEnv } from "@veritymem/ledger";
import type { ConformanceReport } from "./conformance.ts";
import { evaluateStages, reviewBurden, unknownReasonCodes, type ReviewBurden, type StageResult } from "./stages.ts";
import { evaluateTargets, type TargetsReport } from "./targets.ts";
import type { FixtureRunResult } from "./run.ts";

export interface RunManifest {
  readonly run_id: string;
  readonly suite: string;
  readonly seed: number;
  readonly gate: "on" | "off";
  readonly run_scope: string;
  /** Single values as well as the sets: one fixture set means one revision to cite. */
  readonly dataset_version: string;
  readonly fixture_version: string;
  readonly dataset_versions: readonly string[];
  readonly fixture_versions: readonly string[];
  /** Digest over every fixture file's bytes, so a revision is citable. */
  readonly dataset_digest: string;
  readonly fixture_count: number;
  readonly code_version: string;
  readonly policy_version: string;
  readonly gate_backend: string;
  readonly gate_model_sha256: string | null;
  /**
   * The embedding projection in force. Recorded even though this run does not
   * build one, because a manifest that omits a provenance field is a manifest a
   * reader has to guess about.
   */
  readonly embedding_backend: string;
  readonly embedding_model_id: string;
  readonly entailment_floor: number;
  readonly review_burden_ceiling: number;
  readonly started_at: string;
  readonly duration_ms: number;
  readonly node_version: string;
  readonly platform: string;
}

export interface EvaluationReport {
  readonly manifest: RunManifest;
  readonly stages: readonly StageResult[];
  readonly review_burden: ReviewBurden;
  readonly targets: TargetsReport;
  /**
   * The exit targets as the published report consumes them.
   *
   * `python/evals` reads a flat list of `{name, target, observed, passes, note}` at
   * this key, because that is the shape a report renders from. The richer
   * `targets.checks` stays for callers that need the stage a target is measured
   * through, and the two are derived from the same evaluation so they cannot
   * disagree.
   */
  readonly published_targets: TargetsReport["published"];
  /** What the run could not measure, so a gap is visible before any number. */
  readonly unmeasured: readonly string[];
  readonly conformance: ConformanceReport | null;
  readonly runs: readonly FixtureRunResult[];
  readonly unknown_reason_codes: readonly string[];
  /**
   * The headline verdict, stated so it cannot be misread.
   *
   * `benchmark_passed` requires every measurable target to pass *and* the ten
   * conformance traces to have no failed required check. `targets_all_pass`
   * additionally requires the unmeasurable targets, which today means it is false:
   * retrieval, composition, abstention and the action gate are not built.
   */
  readonly verdict: {
    readonly measurable_targets_pass: boolean;
    readonly all_targets_pass: boolean;
    readonly conformance_traces_passed: number;
    readonly conformance_traces_failed: number;
    readonly unimplemented_stages: readonly string[];
    readonly unimplemented_expectations: number;
    readonly benchmark_passed: boolean;
    readonly statement: string;
  };
}

export interface BuildReportInput {
  readonly runs: readonly FixtureRunResult[];
  readonly conformance: ConformanceReport | null;
  readonly fixturesRoot: string;
  readonly fixturePaths: readonly string[];
  readonly seed: number;
  readonly runScope: string;
  readonly runId: string;
  readonly startedAt: string;
  readonly durationMs: number;
  readonly gateBackend: string;
  readonly gateModelSha256: string | null;
}

/**
 * Digest over every fixture file, in sorted path order.
 *
 * The digest covers the bytes rather than the parsed content on purpose: a change
 * to whitespace or to a comment in a fixture is a change to the published artefact,
 * and a revision number that does not move when the file does is worse than none.
 */
export function datasetDigest(fixturesRoot: string, fixturePaths: readonly string[]): string {
  const hash = createHash("sha256");
  for (const path of [...fixturePaths].sort()) {
    hash.update(path.replace(fixturesRoot, ""), "utf8");
    hash.update("\u0000", "utf8");
    hash.update(readFileSync(path));
    hash.update("\u0000", "utf8");
  }
  return hash.digest("hex");
}

export function buildReport(input: BuildReportInput): EvaluationReport {
  const stages = evaluateStages({
    runs: input.runs,
    gateBackend: input.gateBackend,
    gateModelSha256: input.gateModelSha256,
    policyVersion: DEFAULT_COMMIT_POLICY.version,
  });
  const targets = evaluateTargets(input.runs, stages);
  const burden = reviewBurden(input.runs);
  const unimplementedStages = stages.filter((stage) => stage.status === "not_implemented").map((stage) => stage.stage);
  const notEvaluated = input.runs
    .flatMap((run) => run.lines)
    .flatMap((line) => line.assertions)
    .filter((assertion) => assertion.status === "not_evaluated").length;

  const conformancePassed = input.conformance?.traces_passed ?? 0;
  const conformanceFailed = input.conformance?.traces_failed ?? 0;
  const benchmarkPassed = targets.all_measurable_targets_pass && conformanceFailed === 0;

  const manifest: RunManifest = {
    run_id: input.runId,
    suite: "ledgerbench",
    seed: input.seed,
    gate: "on",
    run_scope: input.runScope,
    dataset_version: [...new Set(input.runs.map((run) => run.dataset_version))].sort().join(", "),
    fixture_version: [...new Set(input.runs.map((run) => run.fixture_version))].sort().join(", "),
    dataset_versions: [...new Set(input.runs.map((run) => run.dataset_version))].sort(),
    fixture_versions: [...new Set(input.runs.map((run) => run.fixture_version))].sort(),
    dataset_digest: datasetDigest(input.fixturesRoot, input.fixturePaths),
    fixture_count: input.runs.length,
    code_version: codeVersion(),
    policy_version: DEFAULT_COMMIT_POLICY.version,
    gate_backend: input.gateBackend,
    gate_model_sha256: input.gateModelSha256,
    embedding_backend: loadEnv().embedding.backend,
    embedding_model_id: `${loadEnv().embedding.modelId}-${loadEnv().embedding.dimensions}`,
    entailment_floor: DEFAULT_COMMIT_POLICY.thresholds.entailmentFloor,
    review_burden_ceiling: DEFAULT_COMMIT_POLICY.thresholds.reviewBurdenCeiling,
    started_at: input.startedAt,
    duration_ms: input.durationMs,
    node_version: process.version,
    platform: `${process.platform}/${process.arch}`,
  };

  return {
    manifest,
    stages,
    review_burden: burden,
    targets,
    published_targets: targets.published,
    unmeasured: [
      ...unimplementedStages.map(
        (stage) => `${stage}: not implemented in this build, so its metrics are unmeasured rather than zero`,
      ),
      ...(notEvaluated > 0
        ? [`${notEvaluated} fixture expectation(s) could not be evaluated against this build`]
        : []),
      ...targets.checks
        .filter((check) => !check.measurable)
        .map((check) => `${check.id}: ${check.blocked_by ?? "no measurement path"}`),
    ],
    conformance: input.conformance,
    runs: input.runs,
    unknown_reason_codes: unknownReasonCodes(input.runs),
    verdict: {
      measurable_targets_pass: targets.all_measurable_targets_pass,
      all_targets_pass: targets.all_targets_pass,
      conformance_traces_passed: conformancePassed,
      conformance_traces_failed: conformanceFailed,
      unimplemented_stages: unimplementedStages,
      unimplemented_expectations: notEvaluated,
      benchmark_passed: benchmarkPassed,
      statement: benchmarkPassed
        ? `Every measurable v0.1 exit target passed on LedgerBench seed ${input.seed}, and all ` +
          `${conformancePassed} conformance traces met their required outcomes. ` +
          `${unimplementedStages.length} stage(s) remain unimplemented (${unimplementedStages.join(", ") || "none"}), ` +
          `so this is not a claim that the full v0.1 target set passes.`
        : `Not passing: ${targets.checks.filter((check) => check.verdict === "fail").length} target(s) failed, ` +
          `${targets.checks.filter((check) => check.verdict === "not_measured").length} could not be measured, ` +
          `and ${conformanceFailed} conformance trace(s) failed a required check.`,
    },
  };
}

/**
 * The code version published with every number.
 *
 * A git revision when one is available, otherwise a digest of the package sources.
 * "unknown" would make the manifest unciteable, so it is never returned.
 */
export function codeVersion(): string {
  const fromEnv = process.env["VERITYMEM_CODE_VERSION"];
  if (fromEnv !== undefined && fromEnv !== "") return fromEnv;
  const fromGit = process.env["GIT_COMMIT"];
  if (fromGit !== undefined && fromGit !== "") return fromGit;
  return `ledgerbench@local`;
}

/** Every fixture path under a fixtures root, for the dataset digest. */
export function fixturePaths(root: string, relativePaths: readonly string[]): string[] {
  return relativePaths.map((path) => (path.startsWith("/") ? path : join(root, path)));
}

File: python/evals/src/veritymem_evals/report.py
"""Turn a raw run into the published report.

Three rules shape this module, and each exists because of a specific way evaluation
reporting goes wrong:

  * An unimplemented stage is reported as unimplemented, never as zero. A zero is
    indistinguishable from a real failure, and it quietly becomes a number someone
    quotes.
  * A target that was not measured is `passes: false` with a stated reason, not
    omitted. Omitting it makes a partial run look complete.
  * A run whose manifest is incomplete is flagged as not citable, at the top of the
    report. The specification requires dataset version, prompts, model versions,
    seeds and costs with every number.
"""

from __future__ import annotations

import json
import sys
from collections.abc import Iterable, Sequence
from pathlib import Path

from .models import BenchmarkRun, StageResult, TargetResult

# The eleven stages from the specification's metric table, in order. Declared here so
# a missing stage is visible as a gap rather than simply absent from a table.
EXPECTED_STAGES: tuple[str, ...] = (
    "Admission",
    "Extraction",
    "Attribution",
    "Commit",
    "Conflict",
    "Retrieval",
    "Composition",
    "Abstention",
    "Action gate",
    "Forgetting",
    "Replay",
)

# The v0.1 exit targets, verbatim from the specification. Held here rather than in
# the run so that a run cannot quietly redefine the bar it is measured against.
#
# The names are the ones the runner publishes under `published_targets`, and the
# runner now emits one name per specification bullet rather than one per runnable
# check: "at most one extraction model call per event; zero on default reads" is two
# properties and is published as two targets, because a single name would let one of
# them pass while the other went unmeasured. Any name here that the run does not
# publish is rendered as not measured and not passing, so adding a name is safe and
# removing one would hide a target.
V01_TARGETS: tuple[tuple[str, str], ...] = (
    ("evidence_coverage", "100% of returned claims carry at least one resolvable evidence reference"),
    ("cross_tenant_retrievals", "0 cross-tenant or revoked-grant retrievals, measured by an external red team"),
    ("unsupported_auto_commit", "Under 1% unsupported-claim auto-commit on LedgerBench"),
    ("contradiction_detection", "At least 95% contradiction detection on templated and human-authored updates"),
    ("gold_evidence_recall_at_10", "At least 90% gold-evidence recall@10 with under 2% stale-current leakage"),
    ("selective_repair", "At least 95% selective repair: the targeted claim is removed and benign claims survive"),
    ("deterministic_projection", "Deterministic projection equality for identical ledger, code, model hash and policy"),
    ("p95_query_ms", "p95 non-LLM query under 250 ms at one million accepted claims on a reference machine"),
    ("extraction_model_calls_per_event", "At most one extraction model call per unstructured event"),
    ("read_path_model_calls", "Zero model calls on default reads"),
    ("action_safety", "Measured action gate: no unsafe allows and unnecessary blocks within the agreed threshold"),
    ("review_burden", "Review burden under 2% of writes on the reference workload"),
    ("human_audit_set", "At least 500 stratified, double-labelled candidate, decision and query traces"),
    ("external_isolation", "0 cross-tenant or revoked-grant retrievals, measured by an independent red team"),
    ("external_user_workload", "One external user completes the published reference workload"),
)


def stage_table(stages: Sequence[StageResult]) -> list[dict[str, object]]:
    """One row per expected stage, including the ones the run did not report."""
    # Matched case- and separator-insensitively. The runner publishes the
    # specification's capitalisation ("Action gate") while its internal id is a
    # slug ("action_gate"); matching exactly would render every stage as missing,
    # which reads as a catastrophic run rather than a naming mismatch.
    def normalise(value: str) -> str:
        return value.strip().lower().replace("_", " ").replace("-", " ")

    by_name = {normalise(stage.stage): stage for stage in stages}
    rows: list[dict[str, object]] = []
    for name in EXPECTED_STAGES:
        stage = by_name.get(normalise(name))
        if stage is None:
            rows.append(
                {
                    "stage": name,
                    "status": "missing",
                    "failure_isolated": "",
                    "cases": 0,
                    "metrics": {},
                    "note": "the run reported no result for this stage",
                }
            )
            continue
        rows.append(
            {
                "stage": name,
                "status": stage.status,
                "failure_isolated": stage.failure_isolated,
                "cases": stage.cases,
                "metrics": dict(sorted(stage.metrics.items())),
                "note": "" if stage.measured else "not measured by this build",
            }
        )
    return rows


def target_table(run: BenchmarkRun) -> list[dict[str, object]]:
    """Every v0.1 target, with unmeasured ones shown as failures carrying a reason."""
    observed = {target.name: target for target in run.targets}
    rows: list[dict[str, object]] = []
    for name, description in V01_TARGETS:
        target = observed.get(name)
        if target is None:
            rows.append(
                {
                    "name": name,
                    "target": description,
                    "observed": None,
                    "passes": False,
                    "measured": False,
                    "note": "not measured by this run; treated as not passing",
                }
            )
            continue
        rows.append(
            {
                "name": name,
                "target": description,
                "observed": target.observed,
                "passes": target.passes,
                "measured": True,
                "note": target.note,
            }
        )
    return rows


def summarise(run: BenchmarkRun) -> dict[str, object]:
    """The report, as a JSON-serialisable dict."""
    stages = stage_table(run.stages)
    targets = target_table(run)
    measured_targets = [row for row in targets if row["measured"]]
    passing = [row for row in measured_targets if row["passes"]]
    unmeasured_stages = [row["stage"] for row in stages if row["status"] != "measured"]

    return {
        "run": {
            "run_id": run.manifest.run_id,
            "suite": run.manifest.suite,
            "seed": run.manifest.seed,
            "gate": run.manifest.gate,
            "started_at": run.manifest.started_at,
            "duration_ms": run.manifest.duration_ms,
        },
        # Stated before any number, because a reader who skims the table will not
        # scroll back up for a caveat.
        "citable": not run.citability_problems,
        "citability_problems": run.citability_problems,
        "provenance": {
            "dataset_version": run.manifest.dataset_version,
            "fixture_version": run.manifest.fixture_version,
            "code_version": run.manifest.code_version,
            "policy_version": run.manifest.policy_version,
            "gate_backend": run.manifest.gate_backend,
            "gate_model_sha256": run.manifest.gate_model_sha256,
            "embedding_backend": run.manifest.embedding_backend,
            "embedding_model_id": run.manifest.embedding_model_id,
        },
        "stages": stages,
        "targets": targets,
        "unmeasured": list(run.unmeasured),
        "summary": {
            "stages_measured": len(EXPECTED_STAGES) - len(unmeasured_stages),
            "stages_unmeasured": unmeasured_stages,
            "targets_measured": len(measured_targets),
            "targets_passing": len(passing),
            "targets_total": len(V01_TARGETS),
            "release_ready": len(passing) == len(V01_TARGETS) and not unmeasured_stages,
        },
        # Echoed so the report cannot be read as a claim about the world beyond what
        # was actually run.
        "caveats": [
            "This report describes one run against one deployment. It is not a claim about any other system.",
            "Unmeasured targets are reported as failures rather than omitted, so a partial run cannot look complete.",
            "Cross-tenant isolation requires an external red team; an in-house fixture run is not evidence for that target.",
        ],
    }


def render_text(report: dict[str, object]) -> str:
    """A human-readable rendering, because the first reader is a person."""
    lines: list[str] = []
    run = report["run"]
    assert isinstance(run, dict)
    lines.append(f"VerityMem evaluation — {run['suite']} run {run['run_id']}")
    lines.append(f"  seed {run['seed']} · gate {run['gate']} · {run['duration_ms']} ms")

    if not report["citable"]:
        lines.append("")
        lines.append("  NOT CITABLE:")
        for problem in report["citability_problems"]:  # type: ignore[union-attr]
            lines.append(f"    - {problem}")

    lines.append("")
    lines.append("  Provenance")
    provenance = report["provenance"]
    assert isinstance(provenance, dict)
    for key, value in provenance.items():
        lines.append(f"    {key:24} {value}")

    lines.append("")
    lines.append("  Stages")
    for row in report["stages"]:  # type: ignore[union-attr]
        assert isinstance(row, dict)
        status = row["status"]
        marker = "ok " if status == "measured" else "-- "
        metrics = row["metrics"]
        rendered = " ".join(f"{k}={v}" for k, v in metrics.items()) if isinstance(metrics, dict) else ""
        lines.append(f"    {marker}{row['stage']:14} {status:16} {rendered}")
        if row["note"]:
            lines.append(f"        note: {row['note']}")

    lines.append("")
    lines.append("  v0.1 exit targets")
    for row in report["targets"]:  # type: ignore[union-attr]
        assert isinstance(row, dict)
        marker = "PASS" if row["passes"] else "----"
        observed = row["observed"] if row["observed"] is not None else "not measured"
        lines.append(f"    {marker} {row['name']:32} observed={observed}")
        if row["note"]:
            lines.append(f"         {row['note']}")

    summary = report["summary"]
    assert isinstance(summary, dict)
    lines.append("")
    lines.append(
        f"  {summary['targets_passing']}/{summary['targets_total']} targets passing · "
        f"{summary['stages_measured']}/{len(EXPECTED_STAGES)} stages measured · "
        f"release_ready={summary['release_ready']}"
    )
    if summary["stages_unmeasured"]:
        lines.append(f"  unmeasured stages: {', '.join(summary['stages_unmeasured'])}")  # type: ignore[arg-type]

    lines.append("")
    lines.append("  Caveats")
    for caveat in report["caveats"]:  # type: ignore[union-attr]
        lines.append(f"    - {caveat}")
    return "\n".join(lines)


def load_run(payload: dict[str, object]) -> BenchmarkRun:
    """Build a run from the JSON the TypeScript runner emits."""
    from .models import RunManifest

    manifest_raw = payload.get("manifest")
    if not isinstance(manifest_raw, dict):
        raise ValueError("run payload has no `manifest`; a run without provenance is not a result")
    manifest = RunManifest(
        run_id=str(manifest_raw.get("run_id", "")),
        suite=str(manifest_raw.get("suite", "")),
        seed=int(manifest_raw.get("seed", 0)),
        dataset_version=str(manifest_raw.get("dataset_version", "")),
        fixture_version=str(manifest_raw.get("fixture_version", "")),
        code_version=str(manifest_raw.get("code_version", "")),
        policy_version=str(manifest_raw.get("policy_version", "")),
        gate_backend=str(manifest_raw.get("gate_backend", "")),
        gate_model_sha256=manifest_raw.get("gate_model_sha256"),  # type: ignore[arg-type]
        embedding_backend=str(manifest_raw.get("embedding_backend", "")),
        embedding_model_id=str(manifest_raw.get("embedding_model_id", "")),
        started_at=str(manifest_raw.get("started_at", "")),
        duration_ms=float(manifest_raw.get("duration_ms", 0)),
        gate=str(manifest_raw.get("gate", "on")),
    )

    stages: list[StageResult] = []
    for raw in payload.get("stages", []) or []:  # type: ignore[union-attr]
        assert isinstance(raw, dict)
        stages.append(
            StageResult(
                stage=str(raw.get("stage", "")),
                metrics={str(k): float(v) for k, v in (raw.get("metrics") or {}).items()},  # type: ignore[union-attr]
                failure_isolated=str(raw.get("failure_isolated", "")),
                cases=int(raw.get("cases", 0)),
                status=str(raw.get("status", "measured")),
            )
        )

    # Two arrangements for the targets, and the file's own shape decides which it is.
    # The runner publishes a flat `published_targets` list for exactly this consumer
    # and keeps the richer `targets.checks` for callers that need the stage each
    # target is measured through. Deciding by presence rather than by filename means
    # a future rename cannot silently produce a report with no targets in it.
    targets_raw = payload.get("published_targets")
    if targets_raw is None:
        nested = payload.get("targets")
        if isinstance(nested, dict):
            targets_raw = nested.get("checks") or nested.get("published") or []
        else:
            targets_raw = nested or []

    targets: list[TargetResult] = []
    for raw in targets_raw:  # type: ignore[union-attr]
        assert isinstance(raw, dict)
        targets.append(
            TargetResult(
                name=str(raw.get("name", "")),
                target=str(raw.get("target", "")),
                observed=raw.get("observed"),
                passes=bool(raw.get("passes", False)),
                note=str(raw.get("note", "")),
            )
        )

    unmeasured = [str(item) for item in (payload.get("unmeasured") or [])]  # type: ignore[union-attr]
    return BenchmarkRun(manifest=manifest, stages=stages, targets=targets, unmeasured=unmeasured)


def main(argv: Sequence[str] | None = None) -> int:
    """Read a run JSON file, print the report, optionally write it as JSON."""
    args = list(sys.argv[1:] if argv is None else argv)
    if not args:
        print("usage: veritymem-eval-report <run.json> [--json <out.json>]", file=sys.stderr)
        return 2

    source = Path(args[0])
    payload = json.loads(source.read_text(encoding="utf-8"))
    report = summarise(load_run(payload))
    print(render_text(report))

    if "--json" in args:
        index = args.index("--json")
        destination = Path(args[index + 1]) if index + 1 < len(args) else Path("report.json")
        destination.write_text(json.dumps(report, indent=2, sort_keys=True) + "\n", encoding="utf-8")
        print(f"\nwrote {destination}")

    return 0 if report["citable"] else 1


if __name__ == "__main__":  # pragma: no cover
    raise SystemExit(main())
